library/Notifications/ProvidedHook/Notifications/ObjectsRenderer.php

```diff
@@ -0,0 +1,131 @@
+<?php
+
+/* Icinga Notifications Web | (c) 2024 Icinga GmbH | GPLv2 */
+
+namespace Icinga\Module\Notifications\ProvidedHook\Notifications;
+
+use Generator;
+use Icinga\Module\Icingadb\Common\Database;
+use Icinga\Module\Icingadb\Model\Host;
+use Icinga\Module\Icingadb\Model\Service;
+use Icinga\Module\Icingadb\Widget\ItemList\HostList;
+use Icinga\Module\Icingadb\Widget\ItemList\ServiceList;
+use Icinga\Module\Notifications\Hook\ObjectsRendererHook;
+use ipl\Html\Attributes;
+use ipl\Html\Html;
+use ipl\Html\HtmlElement;
+use ipl\Html\Text;
+use ipl\Html\ValidHtml;
+use ipl\Stdlib\Filter;
+use ipl\Web\Widget\StateBall;
+
+class ObjectsRenderer extends ObjectsRendererHook
+{
+    use Database;
+
+    public function getHtmlForObjectNames(array $objectIdTags): \Generator
+    {
+        $hosts = [];
+        $services = [];
+
+        foreach ($objectIdTags as $tags) {
+            if (isset($tags['service'])) {
+                $services['host'][] = $tags['host'];
+                $services['service'][] = $tags['service'];
+            } else {
+                $hosts[] = $tags['host'];
+            }
+        }
+
+        if (! empty($hosts)) {
+            $hostStates = Host::on($this->getDb())
+                ->filter(Filter::equal('name', $hosts));
+
+            foreach ($hostStates as $host) {
+                $element = new HtmlElement(
+                    'span',
+                    Attributes::create(['class' => 'subject']),
+                    Text::create($host->display_name)
+                );
+
+                yield ['host' => $host->name] => $element;
+            }
+        }
+
+        if (! empty($services)) {
+            $serviceStates = Service::on($this->getDb())
+                ->with(['state', 'host', 'host.state'])
+                ->filter(Filter::all(
+                    Filter::equal('service.name', $services['service']),
+                    Filter::equal('host.name',  $services['host'])
```

Comment:
This results in false positives.

> a on b
> a on c
> b on c

Given that there's also a `b on b`, this will match it